src/api.rs

```diff
@@ -1006,41 +1047,50 @@ impl<T: Pixel> ContextInner<T> {
       FrameType::INTER
     };
     if frame_type == FrameType::KEY {
-      self.gop_output_frameno_start = output_frameno;
-      self.gop_input_frameno_start = input_frameno;
+      *self.gop_output_frameno_start.get_mut(&output_frameno).unwrap() =
+        output_frameno;
+      *self.gop_input_frameno_start.get_mut(&output_frameno).unwrap() =
+        input_frameno;
     }
 
-    let output_frameno_in_gop = output_frameno - self.gop_output_frameno_start;
+    let output_frameno_in_gop =
+      output_frameno - self.gop_output_frameno_start[&output_frameno];
     if output_frameno_in_gop == 0 {
       if output_frameno == 0 {
         let seq = Sequence::new(&self.config);
         // The first frame has no previous frame to copy from
         let fi = FrameInvariants::new_key_frame(
           &FrameInvariants::new(self.config.clone(), seq),
-          self.gop_input_frameno_start
+          self.gop_input_frameno_start[&output_frameno]
         );
-        Ok((fi, true))
+        assert!(!fi.invalid);
+        Ok(fi)
       } else {
         let fi = FrameInvariants::new_key_frame(
           &self.frame_invariants[&(output_frameno - 1)],
-          self.gop_input_frameno_start
+          self.gop_input_frameno_start[&output_frameno]
         );
-        Ok((fi, true))
+        assert!(!fi.invalid);
+        Ok(fi)
       }
     } else {
-      let next_keyframe_input_frameno =
-        self.next_keyframe_input_frameno(self.gop_input_frameno_start, false);
+      let next_keyframe_input_frameno = self.next_keyframe_input_frameno(
+        self.gop_input_frameno_start[&output_frameno],
+        false
+      );
       let fi = FrameInvariants::new_inter_frame(
         &self.frame_invariants[&(output_frameno - 1)],
         &self.inter_cfg,
-        self.gop_input_frameno_start,
+        self.gop_input_frameno_start[&output_frameno],
         output_frameno_in_gop,
         next_keyframe_input_frameno
       );
       if input_frameno >= next_keyframe_input_frameno {
-        return Ok((fi, false));
+        assert!(fi.invalid);
+        return Ok(fi);
       }
-      Ok((fi, true))
+      assert!(!fi.invalid);
```

Comment:
move the assert to the caller, you are checking basically every branch.

Reply:
In the subsequent commit I removed that bool return value so I'm not sure I can do that.

Reply:
wouldn't be simpler to check for `fi.invalid` after you call `set_frame_properties`?

Reply:
It would be, before I removed the `bool` return value. Unless I'm missing something, I believe that now this is the only way to check the correctness of `fi.invalid` according to the branches taken in `build_frame_properties` (and it has already caught one bug during the development of this PR).

Reply:
@lu-zero also note that the fi.invalid condition is opposite for some branches

Reply:
I see, I guess trying to refactor the checks is not really worth it then ^^